src/crypto/tls/common.go

```diff
@@ -674,6 +717,30 @@ type Config struct {
 	// used for debugging.
 	KeyLogWriter io.Writer
 
+	// ClientECHConfigs are the parameters used by the client when it offers the
+	// ECH extension. If a suitable configuration is found and the client
+	// supports TLS 1.3, then ECH will be offered in this handshake.
+	ClientECHConfigs []ech.Config
+
+	// ClientECHDummyConfigs are the parameters used by the client to generate
+	// covertext for the ECH extension. If a suitable configuration is found,
+	// then the client includes a dummy ECH extension to this handshake, which
+	// it expects the server to ignore. This feature is used only if
+	// ClientECHConfigs is not set. Note that sending a dummy ECH extension is
+	// not the same as "offering ECH".
+	ClientECHDummyConfigs []ech.Config
```

Comment:
Which alternative do you prefer:
1. Replace ClientECHDummyConfigs with a flag that, when set, causes the client to send dummy ECH.
2. Remove ClientECHDummyConfigs and send dummy ECH handshake.

Note that an obstacle for (2) is that it might break the openSSL unittests.

Reply:
Another idea:

3. Treat the default (`nil`) value separate from an empty array (`[]ech.Config{}`) type.

Similar to DC, there is a question on whether the feature should be enabled by default. If we would like to enable the feature by default, then the `nil` value (or a `false` value for a boolean type similar to proposal (1)) should be treated as if a dummy ECH is desired.

If it should be disabled by default, then the empty array could be interpreted as an explicit signal whereas `nil` will prevent ECH from being enabled.

Reply:
I confirmed that enabling the feature by default breaks the openSSL tests. That leaves us with a flag that either enables or disables dummy ECH or your suggestion of making a distinction between `nil` from `[]ech.Config{}`.

My strong preference would be a flag. First, this distinction between `nil` and an empty slice isn't made anywhere else in crypto/tls. Second, it seems awkward to have the user pass an `[]ech.Config{}` to trigger a specific behavior.

Reply:
Resolved by adding a flag for enabling ECH covertext.